### ISSUE TITLE
Abnormal high memory usage

### ISSUE THREAD

**Issue by cod3gen** [digibyte/digibyte#144]
I operate Powermining.pw pool and have 4 algorithms from DigiByte available, during last version(6.16.2) i noticed that the DigiByte daemon consumes much more memory than all other daemons i have running. To tune down the memory usage slightly i´ve added "-maxmempool=200 -dbcache=200" parameters to the config. With this config i had a regular memory usage of ~2.3GB pr. daemon instance on DigiByte Core 6.16.2.

I though for sure this would be dealt with on DigiByte Core 6.16.3, but its even worse and i almost regret upgrading. on DigiByte Core 6.16.3 we`re at a stable usage of ~3GB for each daemon instance, with no changes other than upgrading to latest DigiByte Core. 

Will there be any fixes for this in near future? 

Daemons are running on an Ubuntu 16.04.5 LTS instance with 64GB Memory and DigiByte Core RPC client version v6.16.3.0-edb6227

http://prntscr.com/khoo7s

**Comment by minzak** [comment]
Same problem.
Too much memory used, after some time crashed, But not use SWAP, why?
![mem](https://user-images.githubusercontent.com/12154217/46470505-881c7180-c7df-11e8-8731-2d364a89f96c.PNG)

**Comment by red0bear** [comment]
Do you use some special config ?

**Comment by red0bear** [comment]
Do you use some special config ?

**Comment by minzak** [comment]
no, all is default. Still can't limit (

**Comment by piratoskratos** [comment]
My digybited 6.16.5.1 take 2.2 gb of my RAM on my debian. Indeed it's too much when bitcoind take only 0.7 gb.

**Comment by ChillingSilence** [comment]
That's correct, as Bitcoin is only working with 10% of the number of blocks that DigiByte is. Further memory improvements are scheduled for a future release.

**Comment by SmartArray** [comment]
I took myself some time and created a detailed application memory footprint analysis. 

### DigiByte memory analysis summary

`46%` of memory gets allocated in https://github.com/digibyte/digibyte/blob/cd8b35c26325af66b0c45ed997c25baab5bfbb5d/src/validation.cpp#L3026,
which is called from https://github.com/digibyte/digibyte/blob/cd8b35c26325af66b0c45ed997c25baab5bfbb5d/src/txdb.cpp#L263 and the line after that to load the block files from HDD.
It is necessary to keep an index of the blocks (at least the hashes and their related data pointers) in the memory.

`18%` of memory is the overhead necessary to hold the above described data in memory (std::pair)

`18%` of memory is needed to make a sorted index of the height (that is: Integer->BlockPointer): https://github.com/digibyte/digibyte/blob/cd8b35c26325af66b0c45ed997c25baab5bfbb5d/src/validation.cpp#L3940

`18%` is the actual application memory overhead.

### How to fix
This is not something we can fix. In future it will get worse, as more blocks will be added (thus holding more hashes (uint256) plus metadata in memory).

We could however map memory space to the disk instead, which will automatically reduce the memory footprint. This will slightly decrease performance, but nowadays we have SSDs and in future we will even have a lot of NVMe SSDs. 

**This is really important, because in we have yet only mined blocks for 4 years. What happens until year 2035?** 
It would be good if some other core devs can say something about that issue. I am capable to implement my concept. However, we won't be able to solve the issue itself:

Current block height as of 19th November, 2018: ~7.7M
Estimated block height in year 2030: ~30M

Each block consumes at least 32byte for each block + metadata (e.g. memory pointer) + memory management overhead. This results into several GB of memory. With the current implementation it will consume ~10GB in year 2030.

I am aware of Moore's law, but all I want to say is, that devices like the current raspberry pi will slowly run out of support. We could however implement an additional memory efficient mode, that keeps the indices on hard drive.

**Comment by br3nn** [comment]
I look forward to a work around or update to this issue, so i can get my pi up and running for this great coin, my Pi3 post as been closed so :) i will sit and watch this issue thread

**Comment by SmartArray** [comment]
I have already identified and discussed this issue. I think I can easily fix the issue, perhaps this weekend.. The difficult part is QA (testing), it can have side effects we don't know about... Nothing in the source code is documented well enough to guess WHY someone has chosen the design as it is right now

**Comment by br3nn** [comment]
> I have already identified and discussed this issue. I think I can easily fix the issue, perhaps this weekend.. The difficult part is QA (testing), it can have side effects we don't know about... Nothing in the source code is documented well enough to guess WHY someone has chosen the design as it is right now

Mate - :) Your a legend if you can nail this issue, I’m happy to do any testing you want. I have a free pi3, sitting here waitin :P

**Comment by SmartArray** [comment]
Today I implemented the concept I planned during the last week.
Please, everyone, try this: [https://github.com/SmartArray/digibyte/blob/master/doc/build-raspberry.md](url)

Use the code of my repository [https://github.com/SmartArray/digibyte](url)
Please tell me if something does not work. I will fix it as soon as possible

**Comment by ChillingSilence** [comment]
Trying cross-compiling for it from Ubuntu 18.04 x86_64 and it ends:
```
  CXX      libdigibyte_server_a-net.o
  CXX      libdigibyte_server_a-net_processing.o
  CXX      libdigibyte_server_a-noui.o
  CXX      libdigibyte_server_a-outputtype.o
  CXX      libdigibyte_server_a-pow.o
  CXX      libdigibyte_server_a-rest.o
  CXX      script/libdigibyte_server_a-sigcache.o
  CXX      libdigibyte_server_a-shutdown.o
  CXX      libdigibyte_server_a-timedata.o
  CXX      libdigibyte_server_a-torcontrol.o
  CXX      libdigibyte_server_a-txdb.o
  CXX      libdigibyte_server_a-memory_mapping.o
memory_mapping.cpp:12:11: fatal error: sys/syslimits.h: No such file or directory
  #include <sys/syslimits.h>
           ^~~~~~~~~~~~~~~~~
compilation terminated.
Makefile:6995: recipe for target 'libdigibyte_server_a-memory_mapping.o' failed
make[2]: *** [libdigibyte_server_a-memory_mapping.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/home/master/digibyte/src'
Makefile:10781: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/master/digibyte/src'
Makefile:773: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```
Will try directly on the RPi later this evening

**Comment by 0xNtive** [comment]
> I took myself some time and created a detailed application memory footprint analysis.
> 
> ### DigiByte memory analysis summary
> `46%` of memory gets allocated in
> 
> [digibyte/src/validation.cpp](https://github.com/digibyte/digibyte/blob/cd8b35c26325af66b0c45ed997c25baab5bfbb5d/src/validation.cpp#L3026)
> 
> Line 3026 in [cd8b35c](/digibyte/digibyte/commit/cd8b35c26325af66b0c45ed997c25baab5bfbb5d)
> 
>  CBlockIndex* pindexNew = new CBlockIndex(block); 
> ,
> which is called from
> [digibyte/src/txdb.cpp](https://github.com/digibyte/digibyte/blob/cd8b35c26325af66b0c45ed997c25baab5bfbb5d/src/txdb.cpp#L263)
> 
> Line 263 in [cd8b35c](/digibyte/digibyte/commit/cd8b35c26325af66b0c45ed997c25baab5bfbb5d)
> 
>  CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash()); 
> and the line after that to load the block files from HDD.
> It is necessary to keep an index of the blocks (at least the hashes and their related data pointers) in the memory.
> `18%` of memory is the overhead necessary to hold the above described data in memory (std::pair)
> 
> `18%` of memory is needed to make a sorted index of the height (that is: Integer->BlockPointer):
> 
> [digibyte/src/validation.cpp](https://github.com/digibyte/digibyte/blob/cd8b35c26325af66b0c45ed997c25baab5bfbb5d/src/validation.cpp#L3940)
> 
> Line 3940 in [cd8b35c](/digibyte/digibyte/commit/cd8b35c26325af66b0c45ed997c25baab5bfbb5d)
> 
>  vSortedByHeight.reserve(mapBlockIndex.size()); 
> `18%` is the actual application memory overhead.
> 
> ### How to fix
> This is not something we can fix. In future it will get worse, as more blocks will be added (thus holding more hashes (uint256) plus metadata in memory).
> 
> We could however map memory space to the disk instead, which will automatically reduce the memory footprint. This will slightly decrease performance, but nowadays we have SSDs and in future we will even have a lot of NVMe SSDs.
> 
> **This is really important, because in we have yet only mined blocks for 4 years. What happens until year 2035?**
> It would be good if some other core devs can say something about that issue. I am capable to implement my concept. However, we won't be able to solve the issue itself:
> 
> Current block height as of 19th November, 2018: ~7.7M
> Estimated block height in year 2030: ~30M
> 
> Each block consumes at least 32byte for each block + metadata (e.g. memory pointer) + memory management overhead. This results into several GB of memory. With the current implementation it will consume ~10GB in year 2030.
> 
> I am aware of Moore's law, but all I want to say is, that devices like the current raspberry pi will slowly run out of support. We could however implement an additional memory efficient mode, that keeps the indices on hard drive.

How hard is it to implement the memory fix with offloading to hdd?

RAM is getting cheaper and will be more accessible. In this case the industry definitely outperforms our growing mem requirements. I will look into pi options on work memory to see if we can mitigate this with hardware augmentations.

**Comment by kunalbarchha** [comment]
I am not sure whether this issue is still open, closed, or resolved. But I just faced the same issue while setting up full node of DGB. I have 8GB of RAM and dual CPU. At block height 500K + it shut down stating our of memory. I restarted the daemon and it is syncing back from zero. 

I am setting up a pool for DGB, can I prune it to 2048 ? Will pruned blockchain work for mining pool ?

**Comment by ChillingSilence** [comment]
Pruning won't make it utilize less. You should be able to run the node on 4GB still, so you've likely got a bunch of other high-mem usage processes and no swapspace (Or you're running Windows).

**Comment by kunalbarchha** [comment]
2GB of swap, and nothing else on the system. I tried HTOP before starting the digibytedaemon, it was at 2GB usage. after starting DGB daemon, within minutes it jumped to 7GB memory usage. Ubuntu 18.04. Latest DGB source cloned from Master.

**Comment by kunalbarchha** [comment]
No offense, but I have been working with hundreds of different cryptocurrencies since last 3 years, specifically for exchanges where full node is basic requirement. I know the basics. Never have anything else on the server except the node and dependencies.

**Comment by ChillingSilence** [comment]
You should join the DigiByte Support telegram, because there's no way you should be using 2GB on a fresh boot? They'll be able to assist with troubleshooting better than on a closed GitHub issue.

**Comment by kunalbarchha** [comment]
Agreed. There are two app running on the server that will interact with DGB over RPC, but still doesn't change the fact that there's 6GB of RAM and 2GB of SWAP is available.

**Comment by kunalbarchha** [comment]
Anyways, never mind. Will figure it out

**Comment by ChillingSilence** [comment]
No offense but if you were that clever then you'd know that Ubuntu Server only utilizes ~250MB on a fresh boot, not 2GB.
Therefore the issue is not just the Core Wallet, but the other applications that you're using on top and how they're interacting, so you'll want to be looking at those applications you're running, and adding more SWAP.

**Comment by ChillingSilence** [comment]
So the 6GB of RAM remaining and 2GB of swap won't be sufficient depending on the application you're utilizing, basically. Add more swap for starters, it's a really simple test.
Secondly, don't start up the other applications at the same time as DigiByte Core, you'll no doubt see it get fully synchronized and stay up // happy on that amount of system resources.
Finally, if you're already using 2GB on Ubuntu Server, whatever you've got is clearly not "light" on the resources and should be investigated too.

**Comment by kunalbarchha** [comment]
It's a Java app, so it uses 1GB of RAM outright, not more than that. I may reduce that. and give it a try.

**Comment by ChillingSilence** [comment]
Java is a memory hog at the best of times, and likely isn't coping with processing the blocks. Definitely needs more memory.

**Comment by kunalbarchha** [comment]
The app doesn't really process the blocks. It a simple wallet API to create wallet, monitor incoming transaction through walletnofity, and send transaction. That's it. But yes, Java can be a pain sometimes.

**Comment by lurchinms** [comment]
Has this been addressed> 7GB of usage is very large for a coin daemon. I have to run it on its own sever.

**Comment by lurchinms** [comment]
![image](https://user-images.githubusercontent.com/12227665/126487343-07ae27a2-47d2-472d-8cec-21b10e7545a4.png)

**Comment by ChillingSilence** [comment]
This is something that's likely to be incorporated and address in two different ways in version 8, both with the hashmap / btree replacement and also by storing some of that data that would be kept in RAM on the hard drive. It's a side-effect of having so many blocks @lurchinms .

**Comment by Raja4Shekar** [comment]
Hi Guys,
Digibyte uses very high memory (4.2GB RAM) compared to similar blockchains (as shown in attached screen capture) to operate a full node which demands for a high resource system. Kindly have a look into this issue guys.
![image](https://user-images.githubusercontent.com/32691472/194187242-bd8734e7-97ae-42fc-91d7-75b28aef6389.png)
